cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll

```diff
@@ -13,9 +13,18 @@ private import semmle.code.cpp.ir.implementation.raw.internal.InstructionTag
  * Returns `instr` or any instruction used to define `instr`.
  */
 private Instruction getDerivedInstruction(Instruction instr) {
+  // The instruction itself
+  result = instr
+  or
+  // or the GVN definition for the current instruction
   result = valueNumber(instr).getAnInstruction().(StoreInstruction).getSourceValue()
   or
-  result = instr
+  // Special handling for conversions
+  result =
+    getDerivedInstruction(valueNumber(instr).getAnInstruction().(CompareNEInstruction).getLeft())
```

Comment:
You should probably also check that the right-hand side has an `int_value` of 0 to make sure this is _actually_ an int-to-bool conversion.

I also think this deserves a more descriptive comment than what is currently here.

Reply:
Can you explain that a bit? What does a value of 0 or non-zero mean?